File: archive/frontend-server-side-recommendation/recommendation/all-tags.ts
import allTags from "@data/all_tags.json";

export const ALL_TAGS = new Set(
  allTags.map((tag) => tag.trim().toLowerCase()).filter(Boolean),
);

export function normalizeCandidateTag(tag: string) {
  return tag.trim().toLowerCase();
}

export function filterKnownTags(tags: string[]) {
  return tags
    .map(normalizeCandidateTag)
    .filter((tag, index, array) => ALL_TAGS.has(tag) && array.indexOf(tag) === index);
}

File: backend/app/services/recommendation/all_tags.py
from __future__ import annotations

import json
from functools import lru_cache
from pathlib import Path

ROOT_DIR = Path(__file__).resolve().parents[4]
ALL_TAGS_PATH = ROOT_DIR / "data" / "all_tags.json"


@lru_cache(maxsize=1)
def load_all_tags() -> set[str]:
    with ALL_TAGS_PATH.open("r", encoding="utf-8") as file:
        tags = json.load(file)

    return {
        str(tag).strip().lower()
        for tag in tags
        if isinstance(tag, str) and str(tag).strip()
    }


def normalize_candidate_tag(tag: str) -> str:
    return tag.strip().lower()


def filter_known_tags(tags: list[str]) -> list[str]:
    known_tags = load_all_tags()
    result: list[str] = []
    for tag in tags:
        normalized = normalize_candidate_tag(tag)
        if normalized in known_tags and normalized not in result:
            result.append(normalized)
    return result
